File: leetcode/0714.best-time-to-buy-and-sell-stock-with-transaction-fee/solution.cpp
#include "../../common.hpp"
auto fastio{[]() {
  std::ios::sync_with_stdio(0);
  std::cin.tie(0);
  return nullptr;
}()};

// https://leetcode.com/problems/best-time-to-buy-and-sell-stock-with-transaction-fee/
struct Solution {
  static int maxProfit(const vector<int> &prices, int fee) {
    auto it = prices.cbegin(), ite = prices.cend();
    int best0{}, best1{-*it++};
    for (; it != ite; ++it) {
      int nbest0 = std::max(best0, best1 + *it - fee);
      int nbest1 = std::max(best1, best0 - *it);
      best0 = nbest0;
      best1 = nbest1;
    }
    return best0;
  }
};

int main(int argc, char *argv[]) {
  vector<int> v{1, 3, 2, 8, 4, 9};
  int fee = 2;
  std::cout << Solution::maxProfit(v, fee) << "\n";
  return 0;
}

File: leetcode/0714.best-time-to-buy-and-sell-stock-with-transaction-fee/solution.rs
fn main() {
    let n = Solution::max_profit(vec![1, 3, 2, 8, 4, 9], 2);
    println!("{}", n);
}

struct Solution;
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock-with-transaction-fee/
impl Solution {
    pub fn max_profit(prices: Vec<i32>, fee: i32) -> i32 {
        let mut i = prices.into_iter();
        let n = i.next().unwrap_or_default();
        i.fold((0, -n), |(x, y), p| (x.max(y + p - fee), y.max(x - p))).0
    }
}
